src/freenet/io/comm/UdpSocketHandlerImpl.java

```diff
@@ -0,0 +1,472 @@
+package freenet.io.comm;
+
+import com.sun.jna.LastErrorException;
+import com.sun.jna.Native;
+import com.sun.jna.Platform;
+import com.sun.jna.Pointer;
+import com.sun.jna.ptr.IntByReference;
+
+import java.io.FileDescriptor;
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.DatagramSocketImpl;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.SocketException;
+import java.net.SocketTimeoutException;
+import java.util.Random;
+
+import freenet.io.AddressTracker;
+import freenet.io.comm.Peer.LocalAddressException;
+import freenet.node.Node;
+import freenet.node.NodeImpl;
+import freenet.node.PrioRunnable;
+import freenet.support.Logger;
+import freenet.support.io.NativeThread;
+import freenet.support.transport.ip.IPUtil;
+
+public class UdpSocketHandlerImpl implements PrioRunnable, PortForwardSensitiveSocketHandler, UdpSocketHandler {
```

Comment:
On the renamed classes, I do not see the difference. This happened because in one commit there is an interface with the old class name. It's hard to see what the changes were there.

Reply:
I agree that this is problematic. I probably didn't use git correctly when renaming the files. Currently, I see two ways to handle that

(1) You can use a diff tool like meld to compare the *Impl-File with the original file from the next branch. It's not a nice way but since it is "only" seven classes that were renamed, it might be doable. I will avoid such problems in any following commits.

(2) I can try to replace the problematic commits with proper commits. However, I don't know how much effort this would actually require. I would probably create a new branch and cherry-pick the non-problematic commits while replacing the problematic ones.

Which solution would you prefer?  Of course, I'm also open for other solutions.

Reply:
I am a new user and therefore can not recommend something. I can only say that I read PR excluding such files and did not meet anything bad. I also liked the addition of the jacoco tool.

Reply:
Okay, no problem. Thank you for the feedback!

I just looked a bit into solution (2) and my current understanding is that Git's renaming detection does not work well with the steps I performed (renaming as well as modifying the class' java-file and additionally add a new file with the old name that now contains the interface).
As an alternative, instead of renaming each class to "*Impl", I could keep the class names and create an interface "IMyClass" for each class "MyClass". This way, I don't move existing code around, which should make commit diffs cleaner. Furthermore, having an interface called "IMyClass" for each class might generally be a nicer approach than having an "*Impl"-class for each interface.

Reply:
Personally, I prefer the `Impl` option. And on the wiki, I saw a link to the Google code convention as adopted here ..
If you rename the classes and make a commit and then make the interfaces as a separate commit, I think diffs will be visible more cleaner.

Reply:
I think both `IMyClass` and `MyClassImpl` suck.

I prefer an interface with a clean name (_especially_ no Hungarian-type prefixes) and a name for the implementation that alludes to what sets this implementation apart from all the other implementations. I am aware that some (or maybe most?) of the implementations here will probably have no more defining aspect than being the only implementation of the interface in questions but even a `DefaultMyClass` (`MyClass` is a horrible name to illustrate this 😃) is better than a `MyClassImpl`.

Names have meaning! Please don’t waste this opportunity by using some fixed scheme that a machine could apply. :(

Then of course there’s also the question whether we actually need all this if we only want to write tests. Mockito can mock classes, too, and as long as there are accessor methods for things, mocking those is not an issue, either.

Reply:
Besides testing, my intention of creating these interfaces was also to make it easier to replace different parts with other implementations for performance comparison (e.g. to compare different load balancing algorithms or routing algorithms) or maybe even enable re-using most of Freenet's code in a discrete, event-based simulation model). I chose to start with writing tests in order to familiarize myself better with the inner workings of Freenet and the dependencies among the different classes.

Reply:
When I created this pull request, I was not aware of the pull requests created by Toad, which also aim to make it easier to test and perform simulations of Freenet. I guess it does not make sense to further discuss my changes as long as Toad's pull requests have not been dealt with. I am willing to help you with reviewing Toad's changes. However, I consider it unrealistic that a single person can review all the changes of [pull request 576](https://github.com/freenet/fred/pull/576/) anytime soon. If you would be willing to participate in the review, we can split the work. Instead of everyone reviewing single commits, I consider it more realistic that we review the [accumulated changes of a each file](https://github.com/freenet/fred/pull/576/files) and resolve conflicts, if existing.

Reply:
I would like to clarify regarding workflow with PR of not my origin (how to commit resolved conflicts).

Reply:
@mbyrenh unless all logical changesets are always restricted to a single file (possible but unlikely, I guess), that way lies madness.

Reply:
@Bombe Why exactly do you consider this approach problematic? My main concern with reviewing commit-by-commit is that I may spend a lot of time with reviewing and discussing a commit only to discover later on that there is another commit in the PR that changes the code again and renders the previous review and discussion obsolete.

Reply:
@mbyrenh because the unit “commit” has usually been chosen for a very good reason. You don’t want to split a commit into two parts because it touches two files, and you don’t want to lump 2 commits together because they touch the same file. Files and commits are two different axises, and this is valid for both writing and reviewing.

And even if a later commit partially invalidates an earlier one, the history is still important in order to fully understand why the changes were made the way they were. I would never give up that knowledge. 

If you are worried about later commits obsoleting earlier ones, get the commits in order before reviewing, i.e. remove/combine reverts and commits that are like 15 tries of something in a row until it worked.